New HUD Design

ensayia opened the pull request:
The current placeholder HUD is lacking in both design and adhesion to the game's overall goals. I've created a new iteration based on an icon system that follows the game's current status color schema.

Propped above the inventory are now the set of icons for health, hunger, thirst, energy, body temperature, air temperature, and sickness, as well as an indicator for whether temperatures are too hot or too cold. As much as I feel that showing raw stats displaces immersion, I've made it so they can be re-enabled easily under the Exile settings in the main Minetest settings menu. I am open to discussion on whether these should be enabled by default.

The icons are adjusted to fit any aspect ratio or resolution equally, they will look reasonable on any monitor of any size.

Here are some examples of the new content:
![2022-08-21-20:45 11](https://user-images.githubusercontent.com/69833317/185819023-02d76d4f-8fc5-442d-a099-f71efe2ae5a7.png)
![2022-08-21-20:45 45](https://user-images.githubusercontent.com/69833317/185819035-182f137a-20e4-4570-b0b0-3259765bc468.png)
![2022-08-21-20:46 04](https://user-images.githubusercontent.com/69833317/185819042-ead5f0b2-8e68-4cd5-9bfa-0990185b6a6c.png)

I would be humbled if you would take this changes into consideration.

DokimiCU commented:
Hey, that looks good! Very tidy, graphics looks nice.

The HUD has definitely needed improvement. Fixing the screen resolution issues is good too.

I haven't play tested, or checked the code. So these are just some thoughts based on project history:
- The original HUD was similar to this (yours is much better though). The problem we had back then was that people found it confusing, especially beginners. Symbols are obvious only when people already know what they mean. Even "obvious" symbols confuse some people.
- We face a dilemma between immersion (as you say) and giving precise values. Symbols can't show that eating "invertebrate carcass" provides 0.1% food. On the other hand, giving precise values does "make it feel like being the terminator" as someone put it.

Some possibilities:
- maybe have an option to display text above the symbols too. This might help beginners. Text + Symbol + Numbers. People can learn what it all means, then they can turn off the details when they are comfortable.
- maybe rather than having a setting hidden in the menu, have a way to toggle off/on in game by pressing a key (or chat command? Whatever works). People can then toggle through options in game whenever they want extra detail, turn it off again when they want immersion.

Just some thoughts! Overall this might be an improvement on what we have now, even if it can still be improved. 
The more skilled coders will have to judge if the coding is any good :-)

Izzyb1234 commented:
I have this code active on the ex7 development server at exile-biomes.planetofnix.com:30007 if anyone wants to give it a try.  I enabled showing the stat numbers.  Thinking having a chat command to toggle them is needed.  Looks like changing the settings in the client is overridden by the server settings.  I'm not sure how player settings are stored between sessions. would be nice to be able to set your preferences and have it remember them.  too bad there isn't an option to set the size of the fonts :(

ensayia commented:
After learning a bit more about what Minetest is capable of, I reworked the code to use single white images and handle the transparency and coloration in code. This removes the need to use separate images for each stage of urgency and allows easier changes in the future.

As an added bonus, icon transparency is now adjustable in the HUD settings menu. Enjoy!

Izzyb1234 commented:
Found a bug.  If you don't set a value for hud_opacity in minetest.conf, it crashes.  I  changed line 14 to add "or '127'" to the end of the line to resolve in testing.
 
ERROR[Main]: ServerError: AsyncErr: Lua: Runtime error from mod '*builtin*' in callback on_joinplayer(): /home/exile/minetest/games/Exile/mods/health/hud.lua:60: attempt to concatenate upvalue 'hud_opacity' (a nil value)
2022-08-22 23:08:22: ERROR[Main]: stack traceback:
2022-08-22 23:08:22: ERROR[Main]:       /home/exile/minetest/games/Exile/mods/health/hud.lua:60: in function 'setup_hud'
2022-08-22 23:08:22: ERROR[Main]:       /home/exile/minetest/games/Exile/mods/health/hud.lua:191: in function 'func'
2022-08-22 23:08:22: ERROR[Main]:       /home/exile/minetest/builtin/profiler/instrumentation.lua:107: in function </home/exile/minetest/builtin/profiler/instrumentation.lua:100>
2022-08-22 23:08:22: ERROR[Main]:       /home/exile/minetest/builtin/game/register.lua:429: in function </home/exile/minetest/builtin/game/register.lua:415>

Izzyb1234 commented:
Can you add a chat command to adjust transparency in game?  Maybe /hud_transparency?

The game config settings are ignored when connecting to remote games by the looks of it.  

A command to toggle the raw_stats would be nice also.  Maybe /hud_show_stats?

ensayia commented:
> Found a bug. If you don't set a value for hud_opacity in minetest.conf, it crashes. I changed line 14 to add "or '127'" to the end of the line to resolve in testing.

I fixed this as well as adding a default catch value for whether the HUD icons are enabled or not. If I can better figure out the relationship between settingtypes.txt and mineconf.lua I might go with a less hacky fix in the future.

ensayia commented:
This morning I dove into getting chat commands for HUD related things.

The HUD is registered at player join and it doesn't look like HUD elements can be deterministically drawn; if they exist they are drawn and if they don't they are not. Attempting to re-register objects mid-game causes a crash because values being read or written to become nil.  There is also no way to set transparency on text elements of a HUD, color only accepts RGB values but no alpha. A chat command can be used to set the variable in config, but it won't take effect until you log out and back in rendering it no more useful than the settings option.

Looks like this one is off the list unless anyone else knows something I don't.

ensayia commented:
/icon_transparency is now available

It takes a number between 0 and 255, or default which sets it to 127.
non-whole numbers get rounded down, and non-numbers prompt the user to use a number instead.

wizardofgcc commented:
Perhaps to make it more clear how much health/hunger/energy/etc. you have, we could overlay a part of the coloured version of the icon with its height corresponding to the precnetage over a grey one, like furnace fuel burn progress in MTG.
Also, a flashing red outline of the icon once a stat hits a critical level would also be useful if the above is how the icons worked - to tell players the situation is critical even if only a sliver of the icon is left, which might be hard to see (although I'm not sure the flashing animation is possible - didn't find anything in the API docs about formspec animation)

Izzyb1234 commented:
Found a bug.  Need a space after 'to' in "Icon transparency set to50"

Izzyb1234 commented:
I'd do the /show_stats command as a toggle like I did for /hud16

ensayia commented:
Typo fixed.

/show_stats is now a toggle. Good idea.

ensayia commented:
Izzy has been testing this code on his v4 server for a few days now. It's at a good place to be merged. They haven't encountered any further bugs or issues. I think I'll stop here and call this feature complete for the moment.

jeremyshannon commented on mods/health/hud.lua:
```diff
@@ -182,76 +233,116 @@ local function color_envirotemp(v, meta)
 	end
 	if v < stress_low then
 	   overlay = "weather_hud_frost.png"
+	   ttype = "hud_temp_cold"
```
Setting the ttype alongside the overlay is superfluous, we can probably drop lines 255 and 259.

ensayia replied:
Removed them.

jeremyshannon reviewed (commented):
This all seems good. I'm not sure about the clarity of the temperature icons, and the effects icon being synonymous with sickness is a bit off since we have a couple of positive effects, but the former's not a big deal and the latter can be dealt with when the planned "conditions" system goes in -- then we can feed the hud only conditions flagged as negative, and leave the positive ones in the full player effects list.
